### ISSUE TITLE
Computing Time (1080 Ti and Jetson TX2)

### ISSUE THREAD

**Issue by ZeroCool11000** [mrharicot/monodepth#95]
One issue is that mondepth takes really long for a single image of about 0.5 seconds on a 1080 Ti grapic card. Does it depend on the model or something else, i am using the standard script. Is is possible to run monodepth on the nvidia jetson tx2 plattform and are there any empirical value for it ?.

Best regards

**Comment by mrharicot** [comment]
The frame time should be much lower than this for evaluation, what timing are you reporting?
monodepth_simple isn't optimized for speed, it has to build the graph and upload the weights to the GPU which takes the majority of the time.

Moreover in its current state monodepth isn't particularly efficient, the model can be trimmed down significantly without loosing much accuracy.
It is also possible to use smaller images thus speed up both training and testing.

I hope this helps!

**Comment by ZeroCool11000** [comment]
Thx 

More than i thought. 

on 1080 Ti of about 0.5 seconds 
on Nvidia Jetson TX2 about 15 seconds, there are also problems with too less memory.

Maybe also the Code is not so efficent:

def test_simple_video(params):
    print ("\nTest function simple video (monodepth)\n")
    print (args.video_path)
    
    #input video
    cap = cv2.VideoCapture(args.video_path)
    
    #http://answers.opencv.org/question/6805/how-to-get-mjpeg-compression-format-from-webcam/
    
    #save out video
    save = args.video_save
    
    #show video 
    show = args.video_show
    
    if cap.isOpened():
     
        if show:
         cv2.namedWindow("hello simple video(monodepth)", cv2.WINDOW_AUTOSIZE)
        
        #https://stackoverflow.com/questions/25359288/how-to-know-total-number-of-frame-in-a-file-with-cv2-in-python
        frames_count = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))        
        frame_number = 0 
        
        fps = cap.get(cv2.CAP_PROP_FPS)
        size = (int(cap.get(cv2.CAP_PROP_FRAME_WIDTH)),
                int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
        
        # for svaving a video 
        # https://docs.opencv.org/3.0-beta/doc/py_tutorials/py_gui/py_video_display/py_video_display.html
        # Define the codec and create VideoWriter object
        
        output_directory = os.path.dirname(args.video_path)
        output_name = os.path.splitext(os.path.basename(args.video_path))[0]
        
        print(os.path.join(output_directory, "{}_disp.avi".format(output_name)))
        print(output_directory)
        print()
        
        #opencv writer
        fourcc = cv2.VideoWriter_fourcc('M','J','P','G')
        out = cv2.VideoWriter(os.path.join(output_directory, "{}_disp.avi".format(output_name)), fourcc, fps, size)
                
        if not out.isOpened:
         print ("writer open failed")
         return 
        
        while True:
            
            start = time.time()
             
            ret_val, img = cap.read();
            
            print (ret_val)
            
            if not ret_val==True:
             return
            
            """Test function. simple video."""
            
            #resetting graph
            #https://github.com/kratzert/finetune_alexnet_with_tensorflow/issues/8
            tf.reset_default_graph() 

            left  = tf.placeholder(tf.float32, [2, args.input_height, args.input_width, 3])
            model = MonodepthModel(params, "test", left, None)

            #input_image = scipy.misc.imread(args.video_path, mode="RGB")
            input_image=img
            
            original_height, original_width, num_channels = input_image.shape
            input_image = scipy.misc.imresize(input_image, [args.input_height, args.input_width], interp='nearest')
            input_image = input_image.astype(np.float32) / 255
            input_images = np.stack((input_image, np.fliplr(input_image)), 0)
    
            # SESSION
            config = tf.ConfigProto(allow_soft_placement=True)
            sess = tf.Session(config=config)
    
            # SAVER
            train_saver = tf.train.Saver()

            # INIT
            sess.run(tf.global_variables_initializer())
            sess.run(tf.local_variables_initializer())
            coordinator = tf.train.Coordinator()
            threads = tf.train.start_queue_runners(sess=sess, coord=coordinator)

            # RESTORE
            restore_path = args.checkpoint_path.split(".")[0]
            train_saver.restore(sess, restore_path)

            disp = sess.run(model.disp_left_est[0], feed_dict={left: input_images})
            disp_pp = post_process_disparity(disp.squeeze()).astype(np.float32)
            
            #resize image
            disp_to_img = scipy.misc.imresize(disp_pp.squeeze(), [original_height, original_width])
            
            #apply colormap
            #https://www.learnopencv.com/applycolormap-for-pseudocoloring-in-opencv-c-python/
            #https://www.tutorialspoint.com/opencv/opencv_color_maps.htm
            disp_to_img_with_colormap = cv2.applyColorMap(disp_to_img, cv2.COLORMAP_SPRING)
        
            if show:
             cv2.imshow('hello simple video(monodepth)',  disp_to_img_with_colormap)
 
            if save:
              #opencv
              out.write(disp_to_img_with_colormap)
              
            # Press Q on keyboard to stop recording
            if cv2.waitKey(1) & 0xFF == ord('q'):
             break
            
            print("frame number %d of  max frames %d" % (frame_number, frames_count))
            print('done!')
            
            frame_number=frame_number+1
            end = time.time()
            print ("Time per frame: " + str(end - start))
        else:
            print ("camera open failed")

    cv2.destroyAllWindows() 

Best regards

**Comment by mrharicot** [comment]
Your code indeed quite inefficient, you should only create the graph only once!
The only thing which should be in your while True loop should be grabbing the frame from the opencv videoreader, then pushing it through the network (disp = sess.run).
Everything above should not be in the loop.

With a 1080Ti you should get very good performances, I did a similar experiments a few months ago with an old Titan X and got 30fps or so.

I hope this helps!

**Comment by ZeroCool11000** [comment]
Ok thx, yes it helps.
 I am not so familar with tensorflow, but it means i can also do the code more efficent. So moving some parts of the code. 

Thx and best regards

**Comment by vineetkr137** [comment]
is it opossible to crop model for better efficient for higher frame rates..??

**Comment by mrharicot** [comment]
Hi,
The models were trained with 256x512, anything too different from that will not produce reasonable results.
The resnet50 model can be trained with sizes which are multiples of 64, you could then train a 128x256 model for instance, which would theoretically about 4 times faster than the current model.
